### PR TITLE
InterruptableRunnable RunnerInterruptPoint redesign

### PR DESCRIPTION
**Pull request by guliashvili**
There were several inefficiencies in the old version of RunnerInterruptPoint and InterruptableRunnable.

1) RunnerInterruptPoint was throwing the exception when interrupted, however, the exception was always ignored.

2) InterruptableRunnable used the read-write lock, however only write lock was used.

3) InterruptableRunnable InterruptableRunnable, stored almost similar variable stop_, interrupted_.

4) std::atomic<bool> interrupted_ was used with locks, even though it was accessed by default safest access mode memory_order_seq_cst. So no additional cache invalidation was needed.

5) InterruptableRunnable contained code(in method interrupted() and variables bypass_check_, checked) just for testing. Which was slowing down method interrupted().

6) Some more confusing things. notify_all was not needed, as only one thread could be waiting for the conditional variable. RunnerInterruptPoint:: pause(void) looks ambiguous and that's why was not used almost anywhere(Only at one place).

I resolved all these problems by merging InterruptableRunnable and RunnerInterruptPoint into the InterruptableRunnable.

1) No use of the exception.
2) 4) Simple mutex, which is only used for pauseMilli and interrupt. InterruptableRunnable::interrupted is lock-free.
3) Single variable interrupted_.
5) Made InterruptableRunnable::interrupt virtual. Tests override interrupt to make things testable.
6) change to notify_one and removed pause without the specific time.

### CODE REVIEW

File: osquery/dispatcher/tests/dispatcher_tests.cpp [hunk 1]
```diff
@@ -43,8 +43,20 @@ class TestRunnable : public InternalRunnable {
     return i;
   }
 
+  bool interrupted() override {
```
Comment (theopolis) [inline osquery/dispatcher/tests/dispatcher_tests.cpp]: Is it possible to move this into `InternalRunnable`? (prevent code-duplication)

File: osquery/logger/plugins/tests/buffered_tests.cpp [hunk 2]
```diff
@@ -219,17 +232,18 @@ TEST_F(BufferedLogForwarderTests, test_multiple) {
   runner1.check();
   runner2.check();
 }
-
+#include <chrono>
```
Comment (theopolis) [inline osquery/logger/plugins/tests/buffered_tests.cpp]: nit, rm

File: osquery/logger/plugins/tests/buffered_tests.cpp [hunk 3]
```diff
@@ -219,17 +232,18 @@ TEST_F(BufferedLogForwarderTests, test_multiple) {
...
   auto runner = std::make_shared<StrictMock<MockBufferedLogForwarder>>(
       "mock", kLogPeriod);
-  runner->mustRun();
 
   EXPECT_CALL(*runner, send(ElementsAre("foo"), "result"))
       .WillOnce(Return(Status(0)));
   runner->logString("foo");
 
   Dispatcher::addService(runner);
+  // std::this_thread::sleep_for(std::chrono::seconds(2));
```
Comment (theopolis) [inline osquery/logger/plugins/tests/buffered_tests.cpp]: nit, rm

File: include/osquery/dispatcher.h [hunk 4]
```diff
@@ -99,20 +60,13 @@ class InterruptableRunnable {
    * Interruption means resources have been stopped.
    * Non-interruption means no attempt to affect resources has been started.
    */
-  Mutex stopping_;
+  std::mutex stopping_;
```
Comment (theopolis) [inline include/osquery/dispatcher.h]: This is a nit, but I was recommending people to use the `Mutex` alias for consistency.

File: osquery/dispatcher/dispatcher.cpp [hunk 5]
```diff
@@ -20,48 +20,27 @@ namespace osquery {
...
   return interrupted_;
 }
 
 void InterruptableRunnable::pauseMilli(std::chrono::milliseconds milli) {
-  try {
-    point_.pause(milli);
-  } catch (const RunnerInterruptError&) {
-    // The pause request was canceled.
+  if (!interrupted_) {
+    std::unique_lock<std::mutex> lock(stopping_);
```
Comment (theopolis) [inline osquery/dispatcher/dispatcher.cpp]: Another nit, but similar to the `Mutex` alias I was recommending `WriteLock`.

File: include/osquery/dispatcher.h (hunk 4 above)
Comment (guliashvili) [reply include/osquery/dispatcher.h]: Mutex is read-write lock, where both read and write locking mechanisms are a lot more complicated. I don't see any reason to use it if we get advantage of the both read and write lock.
What do you think about that?

File: include/osquery/dispatcher.h [hunk 6]
```diff
@@ -67,52 +39,30 @@ class InterruptableRunnable {
 
  protected:
   /// Allow the runnable to check interruption.
-  bool interrupted();
+  virtual bool interrupted();
```
Comment (akindyakov) [inline include/osquery/dispatcher.h]: I'd guess it should be const method, wouldn't it?

**Comment by guliashvili** [comment]
@theopolis  That's a good idea. I was thinking about adding InternalTestableRunnable in the dispatcher, and that would help in case of the dispatcher_test, however, it's harder for buffered_tests, as MockBufferedLogForwarder is not directly inherited from InternalTestableRunnable, but from BufferedLogForwarder. So will have to create additional BufferedTestableLogForwarder, which is not exactly copy paste but I think, looks uglier.

I will stick with adding it in dispatcher_test, please let me know if you come up to anything better.

File: include/osquery/dispatcher.h (hunk 6 above)
Comment (guliashvili) [reply include/osquery/dispatcher.h]: Not really, in other classes, it's overridden and they are changing the state.(buffered_tests.cpp and dispatcher_tests.cpp)

**Pull request merged**